### DISCUSSION TITLE
`affected` support?

### DISCUSSION THREAD

**Discussion by leonardochaia** [nx-dotnet/nx-dotnet#98] [Q&A] [answered]
Hi! I came here from [here](https://github.com/nrwl/nx/issues/2482#issuecomment-915689752)
I don't really know how NX plugins work so this might be a dumb question.

Can we build only the .NET projects that are affected by a set of changes?

We currently have a monorepo, where we have dotnet and NX separate.
We use a tool I wrote called [dotnet-affected](https://github.com/leonardochaia/dotnet-affected) to build only what is affected.

I would like to use NX for everything, but, for a ~200 project .NET monorepo we really need building what is affected.

For example, given this project structure:

1. Project.Shared
1. Project (depends on .1)
1. Project.Tests (depends on .2)
1. Shared.Tests (depends on .1)

When .2 changes, .3 will be affected so we will build and test .2 and .3. There's no need to build and test .4 since .1 has not changed.

If this plugins support affected, does it work with the NX dependencies?
If so, does that mean that adding a ProjectReference to another project is not enough since there are two places to edit now?

Thank you!!

**Comment by AgentEnder** [comment, marked as answer]
Hey 😄. Everything does work with Nx affected as you would expect. `nx-dotnet` parses the csproj / vbproj / fsproj files to read project references, so you don't need to track the dependencies in 2 places.

**Reply by AgentEnder** [reply]
Here is the portion of the code that performs this: https://github.com/nx-dotnet/nx-dotnet/blob/master/packages/core/src/graph/process-project-graph.ts

**Reply by AgentEnder** [reply]
One thing to note is that it requires each of your .NET projects to have the tag `nx-dotnet`, which we use to not perform extra processing if a project is not relevant. Should keep things speedy, but it something that has to be done. `nx-dotnet` adds this tag automatically for new projects, but if you are migrating existing projects its something to keep in mind.

**Reply by leonardochaia** [reply]
Beautiful!!!!!

I'll give it a shot (can't promise when) and report back.
Thanks bud!
Leo.

**Reply by leonardochaia** [reply]
> One thing to note is that it requires each of your .NET projects to have the tag `nx-dotnet`, which we use to not perform extra processing if a project is not relevant. Should keep things speedy, but it something that has to be done. `nx-dotnet` adds this tag automatically for new projects, but if you are migrating existing projects its something to keep in mind.

I could use `Directory.Build.Props` for that right?

**Reply by leonardochaia** [reply]
Sorry, I'm not sure what 

> tag nx-dotnet

means, but if it's on the docs, don't worry i'll find it

**Reply by leonardochaia** [reply]
> Here is the portion of the code that performs this: https://github.com/nx-dotnet/nx-dotnet/blob/master/packages/core/src/graph/process-project-graph.ts

there's a Visitor for visiting each project? I love this already

**Reply by AgentEnder** [reply]
The tags are set on the nx side, so either in workspace/angular.json or project.json entries that you create for the projects. I need to setup some kind of "import" generator to import projects that are in either appsDir or libsDir but not setup in Nx.

**Reply by leonardochaia** [reply]
Ah! NX tags.
Is there a way to pre-generate them from existing *proj files?

BTW,
> workspace/angular.json or project.json entries that you create for the projects.

We currently use angular.json, which is a gigantic json file and the source of most merge conflicts now days. If there's another way, please point me to it :D

**Reply by AgentEnder** [reply]
In nx 12.4(?)+ you can run `nx g @nrwl/workspace:convert-to-nx-project --all` to switch over to individual project.json files for each project. These files are the same the json for the project from nx.json + angular.json smashed together.

The only prerequisite being that you use workspace version 2, but you can just change the number and run `nx format` to make that transition.

**Reply by AgentEnder** [reply]
There's not currently a way to generate them based on the existing xml project files, thats something an import generator would be able to do. Additionally, you would need the project configuration inside nx for the nx stuff to work. If you are setting that part up already, than its probably easier to add tags while you go

**Reply by leonardochaia** [reply]
> In nx 12.4(?)+ 

Awesome!! we just migrated a few weeks ago so I'll see what I can do.

> There's not currently a way to generate them based on the existing xml project files, thats something an import generator would be able to do. Additionally, you would need the project configuration inside nx for the nx stuff to work. If you are setting that part up already, than its probably easier to add tags while you go

Oh ok, so currently, in order to integrate the plugin with existing .net projects we need to manually add them to the nx files.

Really appreciate your time! Very interesting stuff.

**Reply by AgentEnder** [reply]
Yeah, without adding the projects to nx's config files nx would not know about them. Plugins are just a collection of generators and executors, with some of them also having dep-graph support. 

This isn't just a limitation for community plugins though, one of the cool things about Nx is that even the 1st party nrwl plugins use the same format, and are essentially community plugins authored by nrwl.

**Reply by leonardochaia** [reply]
> Yeah, without adding the projects to nx's config files nx would not know about them. Plugins are just a collection of generators and executors, with some of them also having dep-graph support.

Understood. I will see if I can do some automation when the time comes to try this out.

> one of the cool things about Nx is that even the 1st party nrwl plugins use the same format, and are essentially community plugins authored by nrwl.

One of the many cool things for sure! Makes perfect sense.

I'm definitively interested to try this, so I'll let you know how it goes but I don't know when I'll have some time to get into it :/

Really appreciate you taking the time to answer these questions bro! thanks!! :+1: :+1:

**Reply by AgentEnder** [reply]
@leonardochaia The next release should include @nx-dotnet/core:import-projects to automatically create the configurations so long as the projects have been placed under apps or libs

**Reply by AgentEnder** [reply]
Here are the docs for that: https://www.nx-dotnet.com/docs/core/generators/import-projects

**Reply by leonardochaia** [reply]
Awesome!! that will simplify things for sure.
Hopefully I'll get some time to play with this soon.

Have a great weekend!
Leo.

**Reply by AgentEnder** [reply]
As a quick update in case anyone stumbles upon this: nx-dotnet incudes support for project-inference for several versions now. Part of this is that we no longer require the nx-dotnet tag on projects.
